test_runner/src/main.rs

```diff
@@ -1,3 +1,155 @@
+        let ip_ab = format!("192.168.{}.{}/24", link.0 .1, counter);
+        let ip_ba = format!("192.168.{}.{}/24", link.1 .1, counter);
+        let subnet_a = format!("192.168.{}.0/24", link.0 .1);
+        let subnet_b = format!("192.168.{}.0/24", link.1 .1);
+
+        counter += 1;
+        // create veth to link them
+        let res = KI.run_command(
+            "ip",
+            &[
+                "link", "add", &veth_ab, "type", "veth", "peer", "name", &veth_ba,
+            ],
+        );
+        println!("{:?}", res);
+        // assign each side of the veth to one of the nodes namespaces
+        let res = KI.run_command("ip", &["link", "set", &veth_ab, "netns", &link.0 .0]);
+        println!("{:?}", res);
+        let res = KI.run_command("ip", &["link", "set", &veth_ba, "netns", &link.1 .0]);
+        println!("{:?}", res);
+
+        // add ip addresses on each side
+        let res = KI.run_command(
+            "ip",
+            &[
+                "netns", "exec", &link.0 .0, "ip", "addr", "add", &ip_ab, "dev", &veth_ab,
+            ],
+        );
+        println!("{:?}", res);
+
+        let res = KI.run_command(
+            "ip",
+            &[
+                "netns", "exec", &link.1 .0, "ip", "addr", "add", &ip_ba, "dev", &veth_ba,
+            ],
+        );
+        println!("{:?}", res);
+
+        // bring the interfaces up
+        let res = KI.run_command(
+            "ip",
+            &[
+                "netns", "exec", &link.0 .0, "ip", "link", "set", "dev", &veth_ab, "up",
+            ],
+        );
+        println!("{:?}", res);
+
+        let res = KI.run_command(
+            "ip",
+            &[
+                "netns", "exec", &link.1 .0, "ip", "link", "set", "dev", &veth_ba, "up",
+            ],
+        );
+        println!("{:?}", res);
+
+        //  ip netns exec nC ip route add 192.168.0.0/24 dev veth-nC-nA
+        // add routes to each other's subnets
+        let res = KI.run_command(
+            "ip",
+            &[
+                "netns", "exec", &link.0 .0, "ip", "route", "add", &subnet_b, "dev", &veth_ab,
+            ],
+        );
+        println!("{:?}", res);
+        let res = KI.run_command(
+            "ip",
+            &[
+                "netns", "exec", &link.1 .0, "ip", "route", "add", &subnet_a, "dev", &veth_ba,
+            ],
+        );
+        println!("{:?}", res);
+    }
+
+    Ok(())
+}
+
+/// This struct holds the setup instructions for namespaces
+//#[derive(Debug, Serialize, Deserialize, Clone, Eq, PartialEq)]
+pub struct NamespaceInfo {
+    /// the names of the namespaces
+    pub names: Vec<String>,
+    /// linked nodes written as tuple pairs
+    /// The string is for the namespace name(NOTE: names must be <=4 characters as interfaces cannot be more than 15 char, and we input as veth-{}-{})
+    /// and the u32 is for the subnet on the 3rd octet
+    pub linked: Vec<((String, u32), (String, u32))>,
```

Comment:
The only thing I can think of to add here right now is a function to validate the topology. For example by checking that each link has the appropriate two ends and no names are too long.